File: frontend/src/rpc/client/data/departemen.js
import TDataDepartemenService from '@/rpc/gen/TDataDepartemenService';
import { TUserRole } from '@/rpc/gen/user.user.types_types';
import { TCrudClient } from '@/rpc/client/crud';


class TDataDepartemenClient extends TCrudClient{
	constructor(stores=null){
		super(
			stores, 
			TDataDepartemenService, 
			'/api/data/departemen',
			{
				fetch: TUserRole.ADMIN_UTAMA,
				create: TUserRole.ADMIN_UTAMA,
				delete: TUserRole.SUPER_ADMIN,
			},
			{
				name: TUserRole.SUPER_ADMIN,
				enabled: TUserRole.ADMIN_UTAMA,
			}
		);
	}
}

const dataDepartemenClient = new TDataDepartemenClient();

export { TDataDepartemenClient, dataDepartemenClient }
export default dataDepartemenClient

File: backend/rpc/handler/data/departemen.py
from rpc.gen.data.departemen.services import TDataDepartemenService
from rpc.gen.data.departemen.structs.ttypes import TDepartemen

from rpc.gen.user.user.types.ttypes import TUserRole
from ..crud import TCrudServiceHandler

from models import models

class TDataDepartemenServiceHandler(TDataDepartemenService.Iface, TCrudServiceHandler):
    def __init__(self):
        TDataDepartemenService.Iface.__init__(self)
        TCrudServiceHandler.__init__(
            self,
            "departemen", 
            TDepartemen,
            actions={
                "get": TUserRole.ADMIN_BIASA,
                "create": TUserRole.ADMIN_UTAMA,
                "fetch": TUserRole.ADMIN_BIASA,
                "delete": TUserRole.SUPER_ADMIN,
            },
            setters={
                "name": TUserRole.ADMIN_UTAMA,
                "enabled": TUserRole.ADMIN_UTAMA,
            },
            require_role=False
        )

